### PR TITLE
Improve inferrabililty

### PR DESCRIPTION
**Pull request by timholy**
This makes the methods of this package less vulnerable to invalidation, should improve precompilability, and result in a small runtime performance improvement:

`master`:
```julia
julia> img = rand(RGB{N0f8}, 100, 100);

julia> @btime begin
           TiffImages.save("/tmp/file.tiff", img)
           imgs = TiffImages.load("/tmp/file.tiff")
       end;
  417.934 μs (561 allocations: 110.66 KiB)
```

This branch:
```julia
julia> @btime begin
           TiffImages.save("/tmp/file.tiff", img)
           imgs = TiffImages.load("/tmp/file.tiff")
       end;
  391.745 μs (570 allocations: 112.08 KiB)
```

### CODE REVIEW

**Comment by timholy** [comment]
Some of the individual commits have commit messages that may be worth reading (particularly if you think some of the changes here make the code uglier):
- https://github.com/tlnagy/TiffImages.jl/pull/36/commits/17f0032c291b5846f12404f7dc2c84c3e9f6ed5f
- https://github.com/tlnagy/TiffImages.jl/pull/36/commits/b7fccb7f6237fb3053a17e53e3dcf9be7b90da53

**Comment by timholy** [comment]
CC @IanButterworth

File: src/files.jl
Comment (timholy) [inline src/files.jl:91]: I noticed this first by reading the code, but as I changed method dispatch I sometimes hit it in real life. I think it's pretty important to eliminate this before it becomes a FileIO default.

**Comment by timholy** [comment]
Having "finished" this, were I to do it over again I'd probably add a `S<:Stream` type parameter to `TiffFile`; it would eliminate the ugliest parts of this (the splitting of, e.g., `Base.read` out to `_read`) and might even yield more performance benefits and longer successful inference chains. I'll wait for feedback before making that change, though.

File: src/files.jl
Comment (tlnagy) [reply src/files.jl:91]: Ack. Good catch. I need to just delete this method and figure out a different way to detect strided bilevel TIFFs

**Comment by tlnagy** [comment]
Thanks for taking the time to put this code base through the wringer! I'm sure I understand how `S<:Stream` would help in `TiffFile`? The type definition already specifies io is of type `Stream`?

File: src/files.jl
Comment (tlnagy) [reply src/files.jl:91]: Also, I'm 99% sure replacing io with `tf:TiffFile` would still work and fix the type piracy.

**Comment by timholy** [comment]
Down to
```julia
julia> @btime begin
           TiffImages.save("/tmp/file.tiff", img)
           imgs = TiffImages.load("/tmp/file.tiff")
       end;
  370.517 μs (545 allocations: 111.09 KiB)
```
which is now getting into the territory of considerable improvement. I took the liberty of shifting you from `TiffFile(T, args...)` to `TiffFile{T}(args...)`, mirroring how our Array etc. constructors work these days.

**Review by tlnagy — APPROVED** [review]
I'm happy with these changes now that we parameterize TiffFile on the stream type! Thanks @timholy.

**Comment by IanButterworth** [comment]
Nice! Happy to update https://github.com/tlnagy/TiffImages.jl/pull/35 after this merges

**Comment by tlnagy** [comment]
`TiffImages.jl` master (with this PR) vs ImageMagick for a large TIFF

```julia
julia> using TiffImages
[ Info: Precompiling TiffImages [731e570b-9d59-4bfa-96dc-6df516fadf69]

julia> @time TiffImages.load("/home/tlnagy/Downloads/Gcamp6F_BSA_60px_3_stack.ome.tif");
  2.291251 seconds (3.37 M allocations: 664.546 MiB, 2.18% gc time)

julia> @time TiffImages.load("/home/tlnagy/Downloads/Gcamp6F_BSA_60px_3_stack.ome.tif");
  0.739291 seconds (276.28 k allocations: 504.956 MiB, 2.49% gc time)

julia> using ImageMagick

julia> @time ImageMagick.load("/home/tlnagy/Downloads/Gcamp6F_BSA_60px_3_stack.ome.tif");
  4.011342 seconds (2.70 M allocations: 1.540 GiB, 4.56% gc time)

julia> @time ImageMagick.load("/home/tlnagy/Downloads/Gcamp6F_BSA_60px_3_stack.ome.tif");
  3.180862 seconds (6.72 k allocations: 1.408 GiB, 0.51% gc time)
```

Faster on initial load and way faster on subsequent loads (~5x). Also way less memory usage.

**Comment by timholy** [comment]
Thanks for merging! Note I didn't do anything to solve the piracy issue.

Looking forward to seeing this in action!

**Comment by timholy** [comment]
I am *so* glad to be making progress on moving away from ImageMagick! I'm impressed (and slightly puzzled) that the latency is already lower. That should improve even more with #35.

**Comment by tlnagy** [comment]
> Thanks for merging! Note I didn't do anything to solve the piracy issue.

Fix over in #38. Added another test for this code branch because coverage was a little shallow :sweat_smile:

**Pull request merged**
